sycl/test-e2e/InvokeSimd/Regression/nbarrier_multiple_wg.cpp

```diff
@@ -0,0 +1,214 @@
+#ifdef USE_ACC_PTR
+    local_accessor<int, 1> *LocalAcc,
+#else
+    local_accessor<int, 1> LocalAcc,
+#endif
+    int groupID, int localID, int *o) SYCL_ESIMD_FUNCTION {
+#ifdef USE_ACC_PTR
+  ESIMD_CALLEE_nbarrier<Groups, Threads, Size>(*LocalAcc, groupID, localID, o);
+#else
+  ESIMD_CALLEE_nbarrier<Groups, Threads, Size>(LocalAcc, groupID, localID, o);
+#endif
+}
+
+template <int case_num, unsigned Groups, unsigned Threads, class QueueTY>
+bool test(QueueTY q) {
+  constexpr unsigned Size = VL * Threads * Groups;
+
+  static_assert(Threads > 1, "Threads number must be greater than 1");
+  static_assert(Threads % 2 == 0, "Threads number expect to be even");
+  static_assert(Groups > 1, "Threads number must be greater than 1");
+
+  auto *out = malloc_shared<int>(Size, q);
+  for (int i = 0; i < Size; i++) {
+    out[i] = -1;
+  }
+
+  auto dev = q.get_device();
+  auto deviceSLMSize =
+      dev.template get_info<sycl::info::device::local_mem_size>();
+  std::cout << "Local Memory Size: " << deviceSLMSize << std::endl;
+
+  // The test is going to use Size elements of int type.
+  if (deviceSLMSize < Size * sizeof(int)) {
+    // Report an error - the test needs a fix.
+    std::cerr << "Error: Test needs more SLM memory than device has"
+              << std::endl;
+    return false;
+  }
+
+  try {
+    // workgroups
+    sycl::range<1> GlobalRange{Size};
+    // threads in each group
+    sycl::range<1> LocalRange{Size / Groups};
+    sycl::nd_range<1> Range{GlobalRange * LocalRange, LocalRange};
+
+    auto e = q.submit([&](handler &cgh) {
+      auto LocalAcc = local_accessor<int, 1>(Size, cgh);
+      cgh.parallel_for<KernelID<case_num>>(
+          nd_range<1>(GlobalRange, LocalRange),
+          // This test requires an explicit specification of the subgroup size
+          [=](nd_item<1> item) [[intel::reqd_sub_group_size(VL)]] {
+            sycl::group<1> g = item.get_group();
+            sycl::sub_group sg = item.get_sub_group();
+
+            // Thread local ID in ESIMD context
+            int localID = sg.get_group_linear_id();
+            int groupID = g.get_group_linear_id();
+
+            // SLM init
+            uint32_t slmID = item.get_local_id(0);
+            auto LocalAccCopy = LocalAcc;
+            LocalAccCopy[slmID] = -1;
+            item.barrier();
+
+#ifdef USE_ACC_PTR
+            auto LocalAccArg = uniform{&LocalAccCopy};
+#else
+            auto LocalAccArg = uniform{LocalAccCopy};
+#endif
+
+            invoke_simd(sg, SIMD_CALLEE_nbarrier<Groups, Threads, Size>,
+                        LocalAccArg, uniform{groupID}, uniform{localID},
+                        uniform{out});
+          });
+    });
+    e.wait();
+  } catch (sycl::exception const &e) {
+    std::cout << "SYCL exception caught: " << e.what() << '\n';
+    free(out, q);
+    return false;
+  }
+
+  bool passed = true;
+  for (int i = 0; i < Size; i++) {
+    int etalon = (i / (2 * VL)) * 2;
```

Comment:
also would help to have a comment explaining this

Reply:
I have added some comments to all the places mentioned above and other places throughout the tests. Let me know how I can further improve the readability of the tests.